dialects/index_dialect.py

```diff
@@ -0,0 +1,42 @@
+from xdsl.ir import Attribute, Dialect, OpResult
+from xdsl.irdl import OpAttr, Operand, irdl_op_definition, IRDLOperation
+
+from traits.effects import Pure
+
+
+@irdl_op_definition
+class Add(IRDLOperation, Pure):
+    name = "index.add"
+    lhs: Operand
```

Comment:
```suggestion
    lhs: Annotated[Operand, IndexType]
```
To only allow index types in operands.
Same thing for the other `Operand` and `OpResult`.